### PR TITLE
Evar maps contain econstrs.

### LINKED ISSUE #6310: `Evd.define` is either of the wrong type or shouldn't be used by the current users.
So far `Evd.define` is called by `pretyping` quite a few times, however in all cases the incoming argument is of `EConstr.t` type, whereas the function expects a `Constr.t` type.

This is IMHO a critical bug [could be a design bug or just a use bug], as in all cases the conversion is morally wrong as `Constr.t` terms are not allowed to contain existential variables anymore. See also #6308

### PR DESCRIPTION
**Pull request by SkySkimmer**
We bootstrap the circular evar_map <-> econstr dependency by moving
the internal EConstr.API module to Evd.MiniEConstr. Then we make the
Evd functions use econstr.

Replaces #6500, Fixes #6310.

- [x] Entry added in dev CHANGES.

### CODE REVIEW

**Comment by SkySkimmer** [comment]
https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/377/console

**Review by ejgallego — COMMENTED** [review]
In my view, the right answer for this kind of dependency issues is not specialization, but generalization.

The fact that #6500 could be done is IMHO giving us a strong hint that we are implementing our maps wrong, the maps don't need specific knowledge about the objects they map to, so you really want to make them open.

So I think I much prefer to have a generic `evar_map` that is then later specialized by `Econstr`, vs the current "bootstrapping hack".

That being said, I know the solution @SkySkimmer proposes here is also the favored one by @ppedrot , so I shall remain neutral as to what to do with this PR.

**Review by forestjulien — APPROVED** [review]
this seems ok for the funind part.

**Comment by SkySkimmer** [comment]
```

┌──────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────┐
│                          │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │   mem faults    │
│                          │                         │                                       │                                       │                         │                 │
│             package_name │     NEW     OLD PDIFF   │            NEW            OLD PDIFF   │            NEW            OLD PDIFF   │     NEW     OLD PDIFF   │ NEW OLD PDIFF   │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│      coq-formal-topology │   37.41   37.57 -0.43 % │   100540609336   100632195350 -0.09 % │   125453829319   125641723039 -0.15 % │  483412  483652 -0.05 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-ssreflect │   44.72   44.88 -0.36 % │   122151896882   122147329111 +0.00 % │   142546226677   142720359160 -0.12 % │  537640  537660 -0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│ coq-mathcomp-real_closed │  172.85  173.33 -0.28 % │   478159658548   479115417287 -0.20 % │   700008362898   700460377429 -0.06 % │  837688  843104 -0.64 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-fingroup │   64.55   64.72 -0.26 % │   177534697946   177625459791 -0.05 % │   235934478121   236268530477 -0.14 % │  593304  592968 +0.06 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│    coq-mathcomp-solvable │  201.70  202.17 -0.23 % │   558035088173   558852347218 -0.15 % │   766482889268   767792182374 -0.17 % │  865520  867004 -0.17 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│              coq-bignums │   80.41   80.50 -0.11 % │   221471864620   221527521469 -0.03 % │   285684550186   285991157124 -0.11 % │  528020  527260 +0.14 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│       coq-mathcomp-field │  461.20  461.62 -0.09 % │  1280442111055  1281428467797 -0.08 % │  2060566059039  2061455929807 -0.04 % │  814524  814564 -0.00 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                  coq-vst │ 4010.01 4011.38 -0.03 % │ 11146686844448 11149702488464 -0.03 % │ 14536166777919 14544251942980 -0.06 % │ 2226456 2224628 +0.08 % │   0  42 -100.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│     coq-mathcomp-algebra │  185.32  185.36 -0.02 % │   512337601664   512743036305 -0.08 % │   689411635364   690085997072 -0.10 % │  652292  646404 +0.91 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-corn │ 1566.45 1566.62 -0.01 % │  4333234620578  4336476874804 -0.07 % │  6413831373390  6417277772720 -0.05 % │  899748  903356 -0.40 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│             coq-compcert │  867.14  866.92 +0.03 % │  2402645715552  2401254794058 +0.06 % │  3428348233252  3428994832496 -0.02 % │ 1323296 1322300 +0.08 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│               coq-geocoq │ 3111.67 3110.52 +0.04 % │  8633673168433  8628134945506 +0.06 % │ 13796981051366 13800998647747 -0.03 % │ 1241096 1241544 -0.04 % │   1   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│         coq-fiat-parsers │  705.87  705.38 +0.07 % │  1951918530873  1952940034100 -0.05 % │  2971785485776  2974190782013 -0.08 % │ 3363136 3368112 -0.15 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                 coq-hott │  335.50  335.23 +0.08 % │   931366908218   931792459894 -0.05 % │  1425575082121  1425994449205 -0.03 % │  598988  598948 +0.01 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│            coq-fiat-core │  111.17  111.06 +0.10 % │   308787226160   308884675705 -0.03 % │   376334557459   377031846906 -0.18 % │  501160  502392 -0.25 % │  16   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-odd_order │ 1444.11 1442.42 +0.12 % │  4010796850741  4007029258634 +0.09 % │  6668815844549  6665375754390 +0.05 % │ 1363652 1411836 -3.41 % │   0   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│           coq-coquelicot │   81.44   81.22 +0.27 % │   224472186506   224684911786 -0.09 % │   265324103574   265496035865 -0.06 % │  714076  715364 -0.18 % │   2   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│                coq-flocq │   63.56   63.38 +0.28 % │   174438362892   174171971503 +0.15 % │   211634596821   211653948335 -0.01 % │  662940  663840 -0.14 % │   1   0  +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────┤
│   coq-mathcomp-character │  275.94  274.92 +0.37 % │   765230098517   762666758545 +0.34 % │  1083230961090  1081692111579 +0.14 % │ 1128328 1072848 +5.17 % │   0   0  +nan % │
└──────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────┘


```

**Review by gares — APPROVED** [review]

**Review by ppedrot — APPROVED** [review]
This patch is a bit monolithic w.r.t. the API, but I think there is unluckily no other incremental solution. I personally support this solution over the one advocated #6500. We're not doing category theory here, generalization over a type that is and should only be instantiated once is no progress, that's just over-engineering in my opinion.

**Comment by ejgallego** [comment]
> generalization over a type that is and should only be instantiated once is no progress, that's just over-engineering in my opinion.

Well generalization tends to create _more compact_ code and theorems by avoiding duplication, not the other way around.

Imagine the day `EConstr.t` is subtype of `Constr.t`, then you would need something like a general heap anyways. Or imagine having yet another refinement of Cosntr, that can easily happen depending how you organize things. Anyways, the patch is fine so I won't beat the dead horse more.

**Comment by ejgallego** [comment]
Fixes #6310

**Review by herbelin — COMMENTED** [review]
OK to me.

**Review by ejgallego — APPROVED** [review]
Interesting, quite a few new `Unsafe.to_constr` appear in this patch, this is gonna be fun!

I have nothing to say here other than if @ppedrot agrees I do.

**Comment by ppedrot** [comment]
@mattam82 If you don't comment I'm going to merge this PR when it is rebased...

@SkySkimmer Could you rebase it BTW? Thanks.

**Review by mattam82 — APPROVED** [review]
Cool, I'm happy to see Evd move to EConstr as it should be!

**Comment by ppedrot** [comment]
@SkySkimmer Sorry, you'll have to rebase again before I merge :/

**Comment by SkySkimmer** [comment]
Rebased.
Wait for a test suite job to complete just in case, eg https://travis-ci.org/SkySkimmer/coq/jobs/366051429

**Comment by SkySkimmer** [comment]
Job succeeded.

**Comment by ppedrot** [comment]
I can wait for the full CI maybe...

**Comment by SkySkimmer** [comment]
*shrug* you're the assignee

**Comment by SkySkimmer** [comment]
Did we break equations or is it this PR? https://travis-ci.org/SkySkimmer/coq/jobs/366051439 (abort_on_undefined_evars issue)

**Comment by ejgallego** [comment]
> Did we break equations or is it this PR?

If you are shipping an overlay you need to update it to equations master.

**Comment by SkySkimmer** [comment]
OK done.

**Comment by SkySkimmer** [comment]
I touched the overlay to avoid messing with basic-overlay but previous CI results should be fine.

File: engine/evd.mli
Comment (liyishuai) [inline engine/evd.mli]: Why are these signatures different from those in [evd.ml](https://github.com/coq/coq/pull/7136/files#diff-d5eab4d1ae33866156756e9ba32f0f2eR24)?
> type econstr = constr
type etypes = types

**Pull request merged**
